src/Storages/MergeTree/ReplicatedMergeMutateTaskBase.cpp

```diff
@@ -0,0 +1,225 @@
+                LOG_INFO(log, e.displayText());
+            }
+            else if (e.code() == ErrorCodes::ABORTED)
+            {
+                /// Interrupted merge or downloading a part is not an error.
+                LOG_INFO(log, e.message());
+            }
+            else if (e.code() == ErrorCodes::PART_IS_TEMPORARILY_LOCKED)
+            {
+                /// Part cannot be added temporarily
+                LOG_INFO(log, e.displayText());
+                storage.cleanup_thread.wakeup();
+            }
+            else
+                tryLogCurrentException(log, __PRETTY_FUNCTION__);
+
+            /** This exception will be written to the queue element, and it can be looked up using `system.replication_queue` table.
+                 * The thread that performs this action will sleep a few seconds after the exception.
+                 * See `queue.processEntry` function.
+                 */
+            throw;
+        }
+        catch (...)
+        {
+            tryLogCurrentException(log, __PRETTY_FUNCTION__);
+            throw;
+        }
+
+    }
+    catch (...)
+    {
+        saved_exception = std::current_exception();
+    }
+
+
+    if (saved_exception)
+    {
+        std::lock_guard lock(storage.queue.state_mutex);
+
+        auto & log_entry = selected_entry->log_entry;
+
+        log_entry->exception = saved_exception;
+
+        if (log_entry->type == ReplicatedMergeTreeLogEntryData::MUTATE_PART)
+        {
+            /// Record the exception in the system.mutations table.
+            Int64 result_data_version = MergeTreePartInfo::fromPartName(log_entry->new_part_name, storage.queue.format_version)
+                .getDataVersion();
+            auto source_part_info = MergeTreePartInfo::fromPartName(
+                log_entry->source_parts.at(0), storage.queue.format_version);
+
+            auto in_partition = storage.queue.mutations_by_partition.find(source_part_info.partition_id);
+            if (in_partition != storage.queue.mutations_by_partition.end())
+            {
+                auto mutations_begin_it = in_partition->second.upper_bound(source_part_info.getDataVersion());
+                auto mutations_end_it = in_partition->second.upper_bound(result_data_version);
+                for (auto it = mutations_begin_it; it != mutations_end_it; ++it)
+                {
+                    ReplicatedMergeTreeQueue::MutationStatus & status = *it->second;
+                    status.latest_failed_part = log_entry->source_parts.at(0);
+                    status.latest_failed_part_info = source_part_info;
+                    status.latest_fail_time = time(nullptr);
+                    status.latest_fail_reason = getExceptionMessage(saved_exception, false);
+                }
+            }
+        }
+
+    }
+
+    return false;
+}
+
+
+bool ReplicatedMergeMutateTaskBase::executeImpl()
+{
+    MemoryTrackerThreadSwitcherPtr switcher;
```

Comment:
Maybe std::optional

Reply:
Why? It is RAII class.

Reply:
That's ok, we also call dtor for std::optional :)
Just one needless heap allocation ...